### PR TITLE
Fancy Extruder

### LINKED ISSUE #13276: FancyExtruder
## Reason
We need an extruder that can:

1.  Have variable height elevations
2.  Have variable number of layers per elevation
3.  Can change subdomain IDs in each elevation

## Design
For now - copy in the extrusion code from libMesh and make a new MeshGenerator to do this: `FancyMeshExtruder`

## Impact
Some duplicated code from libMesh.  Ultimately might want to consolidate.

### PR DESCRIPTION
**Pull request by friedmud**
Add an extruder that can have variable a variable height for each elevation, variable number of layers within each elevation and remap subdomain_ids within each elevation

closes #13276

### CODE REVIEW

**Comment by permcody** [comment]
image
I see you've opened a PR. I look forward to commenting on it!

File: framework/src/meshgenerators/FancyExtruderGenerator.C [hunk 1]
```diff
@@ -0,0 +1,547 @@
...
+    _subdomain_swaps(getParam<std::vector<std::vector<subdomain_id_type>>>("subdomain_swaps")),
+    _direction(getParam<Point>("direction")),
+    _has_top_boundary(isParamValid("top_boundary")),
+    _top_boundary(isParamValid("top_boundary") ? getParam<boundary_id_type>("top_boundary") : 0),
+    _has_bottom_boundary(isParamValid("bottom_boundary")),
+    _bottom_boundary(isParamValid("bottom_boundary") ? getParam<boundary_id_type>("bottom_boundary")
+                                                     : 0)
+{
+  // Normalize it
+  _direction /= _direction.norm();
```
Comment (permcody) [inline framework/src/meshgenerators/FancyExtruderGenerator.C]: Can give an FPE!

File: framework/src/meshgenerators/FancyExtruderGenerator.C [hunk 2]
```diff
@@ -0,0 +1,547 @@
...
+
+  const auto num_elevations = _heights.size();
+
+  if (_num_layers.size() != num_elevations)
+    mooseError("The length of 'heights' and 'num_layers' must be the same in ", name());
```
Comment (permcody) [inline framework/src/meshgenerators/FancyExtruderGenerator.C]: `paramError()`

File: framework/src/meshgenerators/FancyExtruderGenerator.C [hunk 3]
```diff
@@ -0,0 +1,547 @@
...
+
+  if (_subdomain_swaps.size() && (_subdomain_swaps.size() != num_elevations))
+    mooseError("If specified, 'subdomain_swaps' must be the same length as 'heights' in ", name());
```
Comment (permcody) [inline framework/src/meshgenerators/FancyExtruderGenerator.C]: paramError()

File: framework/src/meshgenerators/FancyExtruderGenerator.C [hunk 4]
```diff
@@ -0,0 +1,547 @@
...
+  _subdomain_swap_pairs.resize(_subdomain_swaps.size());
+
+  // Reprocess the subdomain swaps to make pairs out of them so they are easier to use
+  for (MooseIndex(_subdomain_swaps) i = 0; i < _subdomain_swaps.size(); i++)
+  {
+    const auto & elevation_swaps = _subdomain_swaps[i];
+    auto & elevation_swap_pairs = _subdomain_swap_pairs[i];
+
+    for (MooseIndex(elevation_swaps) i = 0; i < elevation_swaps.size(); i += 2)
+      elevation_swap_pairs[elevation_swaps[i]] = elevation_swaps[i + 1];
```
Comment (permcody) [inline framework/src/meshgenerators/FancyExtruderGenerator.C]: Need to do sanity checking on your `subdomain_swaps` input, if a user puts in an odd number you'll have an out of bounds access.

Comment (permcody) [reply framework/src/meshgenerators/FancyExtruderGenerator.C]: Also need sanity checking on your swap pairs. If somebody puts in too large of a number, you'll have a separate out of bounds access.

File: framework/src/meshgenerators/FancyExtruderGenerator.C [hunk 5]
```diff
@@ -0,0 +1,547 @@
...
+{
+  // Note: bulk of this code originally from libmesh mesh_modification.C
+  // Original copyright: Copyright (C) 2002-2019 Benjamin S. Kirk, John W. Peterson, Roy H. Stogner
+  // Original license is LGPL so it can be used here.
+
+  auto mesh = _mesh->buildMeshBaseObject();
+
+  std::unique_ptr<MeshBase> input = std::move(_input);
+
+  unsigned int total_num_layers = std::accumulate(_num_layers.begin(), _num_layers.end(), 0);
```
Comment (permcody) [inline framework/src/meshgenerators/FancyExtruderGenerator.C]: This is just bizarre. Is this line of code really in libMesh?! Why not `size()`?

File: framework/src/meshgenerators/FancyExtruderGenerator.C [hunk 6]
```diff
@@ -0,0 +1,547 @@
...
+            // far as I can tell...
+            boundary_info.add_side(new_elem, cast_int<unsigned short>(s + 1), ids_to_copy);
+          }
+          else
+          {
+            // For 1D->2D extrusion, the boundary IDs map as:
+            // Old elem -> New elem
+            // 0        -> 3
+            // 1        -> 1
+            libmesh_assert_less(s, 2);
```
Comment (permcody) [inline framework/src/meshgenerators/FancyExtruderGenerator.C]: I know you copied this code. Do we want to keep the libmesh asserts? I guess they are fine.

**Review by permcody — COMMENTED** [review]
Looks fine except for the complete lack of any input checking. The failure mode will be seg fault for several different kinds of simple input errors. I'd prefer these to be fixed, but if you can't do that now, at least put in a ticket so these cases can be fixed up later.

File: framework/src/meshgenerators/FancyExtruderGenerator.C (hunk 5 above)
Comment (permcody) [reply framework/src/meshgenerators/FancyExtruderGenerator.C]: Ah - nevermind - I see what this is doing now.

File: framework/src/meshgenerators/FancyExtruderGenerator.C [hunk 7]
```diff
@@ -0,0 +1,547 @@
...
+      "The boundary ID to set on the top boundary.  If ommitted one will be generated.");
+
+  params.addParam<boundary_id_type>(
+      "bottom_boundary",
+      "The boundary ID to set on the bottom boundary.  If omitted one will be generated.");
+
+  return params;
+}
+
+FancyExtruderGenerator::FancyExtruderGenerator(const InputParameters & parameters)
```
Comment (YaqiWang) [inline framework/src/meshgenerators/FancyExtruderGenerator.C]: Do we want to call it *Fancy*? Maybe just name it as `ExtruderGenerator` and get rid of the old one. In the future, we may just rename it `Extruder`. I do not think there is anything the old one can do but this new one cannot.

Comment (YaqiWang) [reply framework/src/meshgenerators/FancyExtruderGenerator.C]: I am guessing only moose is using that old extruder generator (not the modifier), so getting rid of it should be possible.

Comment (permcody) [reply framework/src/meshgenerators/FancyExtruderGenerator.C]: Yeah - Fancy is actually a pretty non-descriptive name for this object. Maybe `AdvancedExtruder` or `FlexibleExtruder`, or `VariableLayerExtruder`. If we get rid of the old one than just `Extruder` would be nice too. I suppose we can just default to generating like the old one? It should be difficult to deprecate and remove the existing object.

**Comment by WilkAndy** [comment]
This will be very useful for me, by the way.

File: framework/src/meshgenerators/FancyExtruderGenerator.C (hunk 5 above)
Comment (friedmud) [reply framework/src/meshgenerators/FancyExtruderGenerator.C]: This is my line of code - glad you saw what it does :-)

File: framework/src/meshgenerators/FancyExtruderGenerator.C (hunk 4 above)
Comment (friedmud) [reply framework/src/meshgenerators/FancyExtruderGenerator.C]: ?  I don't see that one.  The size is set by the number of swaps...

BUT: the swaps do need to be _even_

File: framework/src/meshgenerators/FancyExtruderGenerator.C (hunk 7 above)
Comment (friedmud) [reply framework/src/meshgenerators/FancyExtruderGenerator.C]: Well - it's just "fancy".  It's pretty clear that this is most-likely an intermediate name anyway...

**Review by permcody — APPROVED** [review]

**Pull request merged**